### PR TITLE
Adding multi attribute login support when search SearchAllUserStores is enabled

### PR DESCRIPTION
**Pull request by tharakawijekoon**
Related to wso2/product-is#6762
This addition will support multi-attribute login when search SearchAllUserStores is enabled, To enable this you have to set a claim URI or a comma-separated list of claim URIs corresponding to the attributes the user can log in with. The username claim URI does not have to be defined in this list as the user will be searched for with the uid by default. If users are having conflicting claim/attribute values, authentication will fail. Therefore when defining claim URIs as shown below make sure to define claim URIs that have claim/attribute values that are unique across users to use this multi-attribute login.

```
<AuthenticatorConfig name="x509CertificateAuthenticator" enabled="true">
            <Parameter name="AuthenticationEndpoint">https://wso2is.local:9443/x509-certificate-servlet</Parameter>
            <!-- <Parameter name="username">CN</Parameter> -->
            <Parameter name="AlternativeNamesRegex">[a-zA-Z]{7}</Parameter>
            <Parameter name="SearchAllUserStores">true</Parameter>
            <Parameter name="LoginClaimURIs">http://wso2.org/claims/emailaddress,http://wso2.org/claims/mobile</Parameter>
            <!--<Parameter name="setClaimURI">http://wso2.org/claims/identity/userCertificate</Parameter>-->
            <!--<Parameter name="EnforceSelfRegistration">true</Parameter>-->
</AuthenticatorConfig>
```

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 1]
```diff
@@ -276,6 +280,12 @@ private void addOrValidateCertificate(String userName, AuthenticationContext aut
         }
 
         if (isUserCertValid) {
+            try {
+                userName = UserCoreUtil.addDomainToName(userName, getUserStoreDomainName(userName, authenticationContext));
+                UserCoreUtil.setDomainInThreadLocal(getUserStoreDomainName(userName, authenticationContext));
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Shall we assign getUserStoreDomainName(userName, authenticationContext) to a variable and reuse as it is duplicating?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 2]
```diff
@@ -517,5 +527,53 @@ private void addMatchStringsToList(Matcher matcher, Set<String> matches) {
         }
     }
 
+    private String getUserStoreDomainName(String userIdentifier, AuthenticationContext authenticationContext)
+            throws UserStoreException, AuthenticationFailedException {
+        if (Boolean.valueOf(getAuthenticatorConfig().getParameterMap().get(X509CertificateConstants.SEARCH_ALL_USERSTORES))) {
+            String[] filteredUsers = X509CertificateUtil.getUserRealm(userIdentifier).getUserStoreManager().listUsers
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Inorder to increase the readability lets assign X509CertificateUtil.getUserRealm(userIdentifier).getUserStoreManager() to a variable and use it here.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 3]
```diff
@@ -517,5 +527,53 @@ private void addMatchStringsToList(Matcher matcher, Set<String> matches) {
...
+                    (MultitenantUtils.getTenantAwareUsername(userIdentifier), X509CertificateConstants.MAX_ITEM_LIMIT_UNLIMITED);
+            if (filteredUsers.length == 1) {
+                if (log.isDebugEnabled()) {
+                    log.debug("User exists with the user name: " + userIdentifier);
+                }
+                if (filteredUsers[0].indexOf("/") > 0) {
+                    String[] subjectIdentifierSplits = filteredUsers[0].split("/", 2);
+                    return subjectIdentifierSplits[0];
+                } else {
+                    return null;
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Should not we return PRIMARY here?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 4]
```diff
@@ -517,5 +527,53 @@ private void addMatchStringsToList(Matcher matcher, Set<String> matches) {
...
+                }
+            } else if (filteredUsers.length > 1) {
+                authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                        X509CertificateConstants.USERNAME_CONFLICT);
+                throw new AuthenticationFailedException("Conflicting users with user name: " + userIdentifier);
+            } else if (getAuthenticatorConfig().getParameterMap().containsKey(X509CertificateConstants.LOGIN_CLAIM_URIS)) {
+                for (String multiAttributeClaimUri : getAuthenticatorConfig().getParameterMap()
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Shall we assign getAuthenticatorConfig().getParameterMap()
                        .get(X509CertificateConstants.LOGIN_CLAIM_URIS).split(",") to an array first?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 5]
```diff
@@ -517,5 +527,53 @@ private void addMatchStringsToList(Matcher matcher, Set<String> matches) {
...
+                        .get(X509CertificateConstants.LOGIN_CLAIM_URIS).split(",")) {
+                    String[] usersWithClaim = ((AbstractUserStoreManager) X509CertificateUtil.getUserRealm(userIdentifier)
+                            .getUserStoreManager()).getUserList(multiAttributeClaimUri, userIdentifier, null);
+                    if (usersWithClaim.length == 1) {
+                        if (usersWithClaim[0].indexOf("/") > 0) {
+                            String[] subjectIdentifierSplits = usersWithClaim[0].split("/", 2);
+                            return subjectIdentifierSplits[0];
+                        } else {
+                            return null;
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Lets return PRIMARY here.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 6]
```diff
@@ -517,5 +527,53 @@ private void addMatchStringsToList(Matcher matcher, Set<String> matches) {
...
+                    }
+                }
+                throw new AuthenticationFailedException("Unable to find X509 Certificate's user in user store. ");
+            } else {
+                authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                        X509CertificateConstants.USER_NOT_FOUND);
+                throw new AuthenticationFailedException("Unable to find X509 Certificate's user in user store. ");
+            }
+        } else {
+            return null;
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Should not this again PRIMARY

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 7]
```diff
@@ -517,5 +527,53 @@ private void addMatchStringsToList(Matcher matcher, Set<String> matches) {
         }
     }
 
+    private String getUserStoreDomainName(String userIdentifier, AuthenticationContext authenticationContext)
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java:533]: As this primary method is bit long can we split it to few private methods?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 8]
```diff
@@ -517,5 +529,54 @@ private void addMatchStringsToList(Matcher matcher, Set<String> matches) {
...
+        }
+    }
+
+    private String getUserStoreDomainNameHelper(String user) {
+
+        if (user.indexOf("/") > 0) {
+            String[] subjectIdentifierSplits = user.split("/", 2);
+            return subjectIdentifierSplits[0];
+        } else {
+            return X509CertificateConstants.PRIMARY;
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: No need to define a new constant. We can reuse the constant in [1].
[1] https://github.com/wso2/carbon-kernel/blob/4.5.x/core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UserCoreConstants.java#L65

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java (hunk 1 above)
Comment (tharakawijekoon) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Done

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java (hunk 2 above)
Comment (tharakawijekoon) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Done

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java (hunk 3 above)
Comment (tharakawijekoon) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: ok

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 9]
```diff
@@ -335,6 +336,20 @@ private static boolean isUserExists(String userName, AuthenticationContext authe
                 authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
                         X509CertificateConstants.USERNAME_CONFLICT);
                 throw new AuthenticationFailedException("Conflicting users with user name: " + userName);
+            } else if (getX509Parameters().containsKey(X509CertificateConstants.LOGIN_CLAIM_URIS)) {
+                for (String multiAttributeClaimUri : getX509Parameters()
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Assign getX509Parameters()
                        .get(X509CertificateConstants.LOGIN_CLAIM_URIS).split(",") to a variable and use to increase the readability

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateConstants.java [hunk 10]
```diff
@@ -69,5 +69,7 @@ private X509CertificateConstants() {
             + "subjectRDN found for the given regex";
     public static final int MAX_ITEM_LIMIT_UNLIMITED = -1;
     public static final String SEARCH_ALL_USERSTORES = "SearchAllUserStores";
+    public static final String LOGIN_CLAIM_URIS = "LoginClaimURIs";
+    public static final String PRIMARY = "PRIMARY";
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateConstants.java]: No need to define a constant. Re-use [1]

[1] [1] https://github.com/wso2/carbon-kernel/blob/4.5.x/core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/UserCoreConstants.java#L65

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 11]
```diff
@@ -335,6 +336,20 @@ private static boolean isUserExists(String userName, AuthenticationContext authe
...
+                        .get(X509CertificateConstants.LOGIN_CLAIM_URIS).split(",")) {
+                    String[] usersWithClaim = ((AbstractUserStoreManager) X509CertificateUtil.getUserRealm(userName)
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Assign ((AbstractUserStoreManager) X509CertificateUtil.getUserRealm(userName)
                            .getUserStoreManager()) to a variable and re-use

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java [hunk 12]
```diff
@@ -517,5 +529,54 @@ private void addMatchStringsToList(Matcher matcher, Set<String> matches) {
...
+                authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                        X509CertificateConstants.USER_NOT_FOUND);
+                throw new AuthenticationFailedException("Unable to find X509 Certificate's user in user store. ");
+            }
+        } else {
+            return X509CertificateConstants.PRIMARY;
+        }
+    }
+
+    private String getUserStoreDomainNameHelper(String user) {
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java]: Change the parameter name as userIdentifier and change the method name something as getDomainNameByUserIdentifier

**Review by DMHP — APPROVED** [review]

**Pull request merged**